src/man/man8/dump_cat.8
.\" $Revision: 1.17 $
.ds ]W Sun Microsystems
.\" SAM-QFS_notice_begin
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
.\" or https://illumos.org/license/CDDL.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at pkg/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" SAM-QFS_notice_end
.TH dump_cat 8 "27 Sept 2000"
.SH NAME
dump_cat \- Dumps the media changer catalog file in text format
.SH SYNOPSIS
.B /opt/SUNWsamfs/sbin/dump_cat
[
.B \-n
] |
[
.B \-o
]
[
.B \-V
]
.I catalog
.SH AVAILABILITY
.LP
SUNWsamfs
.SH DESCRIPTION
.LP
.B dump_cat
writes a readable form of the \fIcatalog\fP specified on the
command line to standard output.
See
.BR build_cat (8)
for the format of the output.
.SH OPTIONS
.TP 10
.B \-n
Outputs the count of entries in the catalog.
.TP
.B \-o
Lists media that is no longer present in the catalog; i.e., the in-use
flag is not set but there is an entry present.
.TP
.B \-V
Verbose, lists flags and label times as comments.  Lists volume
reservations as comments that may be used to build a ReservedVSNs file.
.SH EXAMPLES
The following is a sample \fBdump_cat\fR listing:
.PP
.ft CO
.nf
# audit_time Wed Dec 31 18:00:00 1969
# version 410  count 32 mediatype 
# Index  VSN     Barcode  Type   PTOC Access Capacity  Space Status    Sector Label time  Slot Part Modification time Mount time Reserved Time Archive-Set/Owner/File System Volume Location
#   ---status---  ---label time----  --last mod time--  ----mount time---
#
  0 004974 004974 sg       0    3 19915760 19915760 0x6a000200 131072 0x3f
8aec0f     0   0          0 0x3fdf5a79          0 // NO_INFORMATION
#   -il-o-b-----  10/13/03 13:16:47  12/31/69 18:00:00  12/16/03 13:18:17
  1 000120 000120 sg     0x4    4 19915760 19915760 0x6a000200 262144 0x3f
d0ae8e     1   0 0x3fdf51c0 0x3fdf51be          0 // "1440 Northland Drive, Shelf 15"
#   -il-o-b-----  12/05/03 10:13:02  12/16/03 12:41:04  12/16/03 12:41:02
 14 000139 000139 sg    0x20    2 19915760 19915760 0x7a000200 131072 0x3f
8aec20    14   0 0x3fdf51ea 0x3fdf51e7          0 // "vsn 000139 arset.0 from /samfs1"
#   -ilEo-b-----  10/13/03 13:17:04  12/16/03 12:41:46  12/16/03 12:41:43
 15 700178 700178 sg   0x804    3 19915760 19382920 0x7a000a00 262144 0x3f
d4f746    15   0 0x3fdf5317 0x3fdf5213          0 // NO_INFORMATION
.PP
.ft
.fi
.SH SEE ALSO
.BR build_cat (8),
.BR sam-robotsd (8).


src/catalog/dump_cat/dump_cat.c
/*
 * dump_cat.c - dump a SAM catalog file in ASCII text.
 */

/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or https://illumos.org/license/CDDL.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

#pragma ident "$Revision: 1.27 $"

static char *_SrcFile = __FILE__; /* Using __FILE__ makes duplicate strings */

/* ANSI C headers. */
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* POSIX headers. */
#include <sys/types.h>
#include <unistd.h>
#include <fcntl.h>

/* SAM-FS headers. */
#define DEC_INIT
#include "sam/types.h"
#include "aml/catalog.h"
#include "aml/catlib.h"
#include "sam/custmsg.h"
#include "sam/exit.h"
#include "sam/lib.h"
#include "sam/sam_malloc.h"

/* Local includes. */
#include "catlib_priv.h"

#if defined(lint)
#include "sam/lint.h"
#endif /* defined(lint) */

/* Private functions. */
static void PrintCatalogEntry(int index, struct CatalogEntry *ce);
static int cmp_slot(const void *p1, const void *p2);

/* Private data. */
static boolean_t verbose = FALSE;

int
main(
	int argc,
	char **argv)
{
	struct CatalogHdr *ch;
	struct CatalogEntry **ce_sort;
	struct CatalogEntry ce_blank;
	size_t	size;
	boolean_t show_old = FALSE;
	boolean_t show_count = FALSE;
	void	*mp;
	char	*cf_name;
	char	*program_name;
	int	c;
	int	count;
	int	errors = 0;		/* Errors encountered */
	int	index;
	int	last_slot;
	int	n;
	int	version;		/* Catalog file version number */
	time_t	audit_time;

	program_name = *argv;

	/*
	 * Process arguments.
	 */
	while ((c = getopt(argc, argv, "Von")) != EOF) {
		switch (c) {
		case 'V':
			verbose = TRUE;
			break;
		case 'n':
			show_count = TRUE;
			break;
		case 'o':
			show_old = TRUE;
			break;
		default:
			errors++;
			break;
		}
	}

	/*
	 * Must have only one argument left for the file name.
	 */
	if ((argc - optind) > 1) {
		fprintf(stderr, "%s: %s\n", program_name, GetCustMsg(4600));
		errors++;
	}
	if ((errors != 0) || ((argc - optind) < 1) ||
	    (show_count && (verbose || show_old))) {
		fprintf(stderr,
		    "%s %s [-n] | [-o] [-V] catalog_file\n",
		    GetCustMsg(4601), program_name);
		exit(EXIT_USAGE);
	}

	/*
	 * The next argument is the catalog file.
	 */
	cf_name = argv[optind];

	version = CatalogMmapCatfile(cf_name, O_RDONLY, &mp, &size, NULL);

	/*
	 * Display the count of entries.
	 */
	ch = mp;
	if (show_count) {
		printf("%d\n", ch->ChNumofEntries);
		exit(EXIT_SUCCESS);
	}

	/*
	 * Print the catalog table header.
	 * build_cat ignores lines beginning with "#".
	 */
	ch = mp;

        audit_time = ch->ChAuditTime;
	/* ctime adds \n */
	printf("# audit_time %s", ctime(&audit_time));
	printf("# version %d  count %d mediatype %s\n",
	    ch->ChVersion, ch->ChNumofEntries, ch->ChMediaType);

	if (version != CF_VERSION) {
		fprintf(stderr, GetCustMsg(18450), version, CF_VERSION);
		fprintf(stderr, "\n");
		exit(EXIT_FAILURE);
	}

	printf("%s\n", GetCustMsg(18451));

/*  Print the mid now, just for debugging.  build_cat ignores it. hgm */
/*
 * # Index  VSN  Barcode Type PTOC Access Capacity Space Status Sector
 * Label time Slot Part Modification time Mount time Reserved time
 * Archive-Set/Owner/File-System
 */
	if (verbose) {
		printf("%s\n", GetCustMsg(18452));
/*
 * #      ---status---  ---label time----  --last mod time--  ----mount time---
 */
	}
	printf("#\n");

	/*
	 * Print all the catalog entries in slot order.
	 */
	size = ch->ChNumofEntries * sizeof (struct CatalogEntry *);
	SamMalloc(ce_sort, size);
	count = 0;
	for (n = 0; n < ch->ChNumofEntries; n++) {
		struct CatalogEntry *ce;

		ce = &ch->ChTable[n];
		if (ce->CeStatus & CES_inuse) {
			ce_sort[count++] = &ch->ChTable[n];
		}
	}
	qsort(ce_sort, count, sizeof (struct CatalogEntry *), cmp_slot);
	memset(&ce_blank, 0, sizeof (ce_blank));

	last_slot = 0;
	index = 0;
	for (n = 0; n < count; n++) {
		struct CatalogEntry *ce;

		ce = ce_sort[n];
		while (last_slot+1 < ce->CeSlot) {
			if (show_old) {
				PrintCatalogEntry(index, &ce_blank);
			}
			index++;
			if (strcmp(ch->ChMediaType, "mo") == 0) {
				if (show_old) {
					PrintCatalogEntry(index, &ce_blank);
				}
				index++;
			}
			last_slot++;

		}
		PrintCatalogEntry(index, ce);
		last_slot = ce->CeSlot;
		index++;
	}

	return (EXIT_SUCCESS);
}


/*
 * Print a catalog entry.
 */
void
PrintCatalogEntry(
	int index,
	struct CatalogEntry *ce)
{
	int blanks = FALSE;
	char TmpBarCode[BARCODE_LEN+3];
	char TmpVolInfo[VOLINFO_LEN+3];

	/*
	 * Search the barcode for blanks. If present, place
	 * quotes around the barcode so this output can be used by build_cat.
	 */
	if (strchr(ce->CeBarCode, ' ') != NULL) {
		blanks = TRUE;
	}

	if ((*ce->CeBarCode != '\0') && (ce->CeStatus & CES_bar_code)) {
		if (blanks) {
			sprintf(TmpBarCode, "\"%s\"", ce->CeBarCode);
		} else {
			sprintf(TmpBarCode, "%s", ce->CeBarCode);
		}
	} else {
		sprintf(TmpBarCode, "%s", NO_BAR_CODE);
	}

	/*
	 * Now do the same for the volume information field.
	 */
	blanks = FALSE;
	if (strchr(ce->CeVolInfo, ' ') != NULL) {
		blanks = TRUE;
	}

	if (*ce->CeVolInfo != '\0') {
		if (blanks) {
			sprintf(TmpVolInfo, "\"%s\"", ce->CeVolInfo);
		} else {
			sprintf(TmpVolInfo, "%s", ce->CeVolInfo);
		}
	} else {
		sprintf(TmpVolInfo, "%s", NO_INFORMATION);
	}

	printf(
	    "%c%5d %6s %6s %2s %#10llx %4d %7lld %7lld %#10x %5d %#10x %5d "
	    "%3d %#10x %#10x %#10x %s/%s/%s %s\n",
	    (ce->CeStatus & CES_inuse) ? ' ' : '#',
	    index,
	    (*ce->CeVsn != '\0') ? ce->CeVsn : "?",
	    TmpBarCode,
	    (*ce->CeMtype != '\0') ? ce->CeMtype : "??",
	    ce->m.CePtocFwa,
	    ce->CeAccess,
	    ce->CeCapacity,
	    ce->CeSpace,
	    *(int *)&ce->CeStatus,
	    ce->CeBlockSize,
	    ce->CeLabelTime,
	    ce->CeSlot,
	    ce->CePart,
	    ce->CeModTime,
	    ce->CeMountTime,
	    ce->r.CerTime,
	    ce->r.CerAsname,
	    ce->r.CerOwner,
	    ce->r.CerFsname,
	    TmpVolInfo);

	if (verbose) {

		/*
		 * Print status and times.
		 * build-cat ignores these since they print as comments.
		 */
		time_t	Time;
		struct tm *tm;
		char	buf[80];

		printf("#      %s", CatalogStatusToStr(ce->CeStatus, buf));
		Time = ce->CeLabelTime;
		tm = localtime(&Time);
		strftime(buf, sizeof (buf), "%x %X", tm);
		printf("  %s", buf);
		Time = ce->CeModTime;
		tm = localtime(&Time);
		strftime(buf, sizeof (buf), "%x %X", tm);
		printf("  %s", buf);
		Time = ce->CeMountTime;
		tm = localtime(&Time);
		strftime(buf, sizeof (buf), "%x %X", tm);
		printf("  %s\n", buf);

		/*
		 * ReservedVSNs entry.
		 */
		if (ce->r.CerTime != 0) {
			Time = ce->r.CerTime;
			tm = localtime(&Time);
			strftime(buf, sizeof (buf), "%Y/%m/%d %H:%M:%S", tm);
			printf("#R %s %s %s/%s/%s %s\n",
			    ce->CeMtype,
			    ce->CeVsn,
			    ce->r.CerAsname,
			    ce->r.CerOwner,
			    ce->r.CerFsname,
			    buf);
		}
	}
}


/*
 * Compare slot numbers.
 */
static int
cmp_slot(
	const void *p1,
	const void *p2)
{
	struct CatalogEntry *ce1 = *(struct CatalogEntry **)p1;
	struct CatalogEntry *ce2 = *(struct CatalogEntry **)p2;

	if (ce1->CeSlot < ce2->CeSlot) {
		return (-1);
	}
	if (ce1->CeSlot > ce2->CeSlot) {
		return (-1);
	}
	if (ce1->CePart < ce2->CePart) {
		return (-1);
	}
	if (ce1->CePart > ce2->CePart) {
		return (1);
	}
	return (0);
}
